### PR TITLE
A pull request for (mainly) homomorphic adder and multiplier

### PR DESCRIPTION
**Pull request by howpenghui**
The main work:
Implemented Homomorphic shifters (left and right), adder (with carry) and multiplier (integer multiplication in base 2)
Implemented + (with carry) and * (integer multiplication) operators in BitVector
Tested the above (in the gtest framework)
Added local tests for the above, in chrono, updated Matt with the new time stats (in email)

TODO:
Reduce public key size so that the execution doesn't segfault due to excessively large public key
Clean code (e.g. there are multiple things that can be sped up using operations such as memmove, good row operations, etc)
Compiler into JavaScript so that we can update the runtime in JS as well
Update the time stats in the paper once the above are done (otherwise, how are we supposed to deal with the 128-bit case when there's segfault?)

### CODE REVIEW

File: krypto-lib/src/main/cpp/PublicKey.h [hunk 1]
```diff
@@ -54,15 +56,38 @@ class PublicKey{
...
+			result = homomorphicADD(result, homomorphicAND(homomorphicLMM(_lc, shiftedX), y));
+		}
+		return result;
+	}
+//*/
 private:
 	const BitMatrix<2*N, 4*N> _ls;
 	const BitMatrix<2*N, 4*N> _rs;
+	const BitMatrix<2*N, 4*N> _lc;
+	//const BitMatrix<2*N, 4*N> _rc;
```
Comment (geekbeast) [inline krypto-lib/src/main/cpp/PublicKey.h]: Don't leave commented code in the final pull request.

File: krypto-lib/src/main/cpp/BitVector.h [hunk 2]
```diff
@@ -444,7 +510,7 @@ class BitVector {
      */
     const BitVector<NUM_BITS> rightShift(unsigned int n){
         BitVector<NUM_BITS> result = BitVector<NUM_BITS>::zeroVector();
-        for(unsigned int i = n; i < NUM_BITS; ++i)
+        for(unsigned int i = NUM_BITS; i >= n; --i)
```
Comment (geekbeast) [inline krypto-lib/src/main/cpp/BitVector.h]: I don't think this is used by any performance critical sections, but we should have a TODO to make this more efficient.

**Comment by geekbeast** [comment]
Overall looks good to me. Tagging @UnsungHero97 to make sure the KryptnosticClient.js that gets generated after this pull requests gets the right version number.

**Comment by howpenghui** [comment]
@geekbeast @UnsungHero97 I have made the change, i.e. removed the commented lines.
Besides, I agree with the comment about inefficient implementation. It can be fixed by using memmove instead of for loop.

**Pull request merged**